packages/core/src/tools/instrumentMethod.ts

```diff
@@ -2,67 +2,80 @@ import { setTimeout } from './timer'
+}
+
+export type PostCallCallback<TARGET extends { [key: string]: any }, METHOD extends keyof TARGET> = (
+  result: ReturnType<TARGET[METHOD]>
+) => void
+
+export function instrumentMethod<TARGET extends { [key: string]: any }, METHOD extends keyof TARGET>(
+  targetPrototype: TARGET,
   method: METHOD,
   instrumentationFactory: (
-    original: OBJECT[METHOD]
-  ) => (this: OBJECT, ...args: Parameters<OBJECT[METHOD]>) => ReturnType<OBJECT[METHOD]>
+    original: TARGET[METHOD]
+  ) => (this: TARGET, ...args: Parameters<TARGET[METHOD]>) => ReturnType<TARGET[METHOD]>
 ) {
-  const original = object[method]
+  const original = targetPrototype[method]
 
   let instrumentation = instrumentationFactory(original)
 
-  const instrumentationWrapper = function (this: OBJECT): ReturnType<OBJECT[METHOD]> | undefined {
+  const instrumentationWrapper = function (this: TARGET): ReturnType<TARGET[METHOD]> | undefined {
     if (typeof instrumentation !== 'function') {
       return undefined
     }
     // eslint-disable-next-line @typescript-eslint/no-unsafe-return
-    return instrumentation.apply(this, arguments as unknown as Parameters<OBJECT[METHOD]>)
+    return instrumentation.apply(this, arguments as unknown as Parameters<TARGET[METHOD]>)
   }
-  object[method] = instrumentationWrapper as OBJECT[METHOD]
+  targetPrototype[method] = instrumentationWrapper as TARGET[METHOD]
 
   return {
     stop: () => {
-      if (object[method] === instrumentationWrapper) {
-        object[method] = original
+      if (targetPrototype[method] === instrumentationWrapper) {
+        targetPrototype[method] = original
       } else {
         instrumentation = original
       }
     },
   }
 }
 
-export function instrumentMethodAndCallOriginal<OBJECT extends { [key: string]: any }, METHOD extends keyof OBJECT>(
-  object: OBJECT,
+export function instrumentMethodAndCallOriginal<TARGET extends { [key: string]: any }, METHOD extends keyof TARGET>(
+  targetPrototype: TARGET,
   method: METHOD,
-  {
-    before,
-    after,
-  }: {
-    before?: (this: OBJECT, ...args: Parameters<OBJECT[METHOD]>) => void
-    after?: (this: OBJECT, ...args: Parameters<OBJECT[METHOD]>) => void
-  }
+  onCall: (this: null, callInfos: InstrumentedMethodCall<TARGET, METHOD>) => void
 ) {
   return instrumentMethod(
-    object,
+    targetPrototype,
     method,
     (original) =>
       function () {
-        const args = arguments as unknown as Parameters<OBJECT[METHOD]>
+        const parameters = arguments as unknown as Parameters<TARGET[METHOD]>
         let result
 
-        if (before) {
-          callMonitored(before, this, args)
-        }
+        let postCallCallback: PostCallCallback<TARGET, METHOD> | undefined
+
+        callMonitored(onCall, null, [
+          {
+            target: this,
+            parameters,
+            onPostCall: (callback) => {
+              postCallCallback = callback
+            },
+          },
+        ])
```

Comment:
Yes we do, so we can use it in `fetchObservable`: in that module we would need to use both the `before` and `after`, and transmit some context between the two. This is why I added a way to register a "post call" callback, so we can just use the instrumentation callback scope to keep the context.

An alternative could be to allow storing the context somehow (maybe in the `InstrumentedMethodCall` object), but typings would be a bit more hazarduous (ex: it would require more casts [like we do in request collections](https://github.com/DataDog/browser-sdk/blob/b45ff9b8c20f9af118e1a4320cea6e7461769df8/packages/rum-core/src/domain/requestCollection.ts#L113)).

> it requires to go look into the instrument method implementation to understand how things will work

I can certainly add more documentation if you think it's needed.

Reply:
could it work with having both a "pre call" callback and a "post call" callback in the instrumentation callback?